src/state/apis/zapper/zapperApi.ts

```diff
@@ -240,7 +303,298 @@ export const zapper = createApi({
+                  const apy = bnOrZero(asset.dataProps?.apy).div(100).toString()
+                  const tvl = bnOrZero(asset.dataProps?.liquidity).toString()
+                  const icon = asset.displayProps?.images?.[0] ?? ''
+                  const name = asset.displayProps?.label ?? ''
+                  const groupId = asset.groupId
+                  const type = asset.type
+                  const assetId = toAssetId({
+                    chainId:
+                      ZAPPER_NETWORKS_TO_CHAIN_ID_MAP[asset.network as SupportedZapperNetwork],
+                    assetNamespace: 'erc20', // TODO(gomes): this might break for native assets, is that a thing?,
+                    assetReference: asset.address,
+                  })
+
+                  const opportunityId: StakingId = `${asset.address}#${asset.key}`
+
+                  if (!acc.metadataByProvider[appName]) {
+                    acc.metadataByProvider[appName] = {
+                      provider: appName,
+                      icon: appImage,
+                      color: '#000000', // TODO
+                      url:
+                        appId && maybeZapperV2AppsData.data
+                          ? maybeZapperV2AppsData.data[appId]?.url
+                          : '',
+                    }
+                  }
+                  // TODO(gomes): ensure this works with the heuristics of ContractPosition vs AppToken
+                  const underlyingAssetIds = asset.tokens.map(token => {
+                    const underlyingAssetId = toAssetId({
+                      chainId:
+                        ZAPPER_NETWORKS_TO_CHAIN_ID_MAP[token.network as SupportedZapperNetwork],
+                      assetNamespace: 'erc20', // TODO(gomes): this might break for native assets, is that a thing?,
+                      assetReference: token.address,
+                    })
+                    return underlyingAssetId
+                  }) as unknown as AssetIdsTuple
+
+                  // An "AppToken" is a position represented a token, thus it is what we call an underlyingAssetId i.e UNI LP AssetId
+                  const underlyingAssetId =
+                    asset.type === 'app-token' ? assetId : underlyingAssetIds[0]!
+
+                  // Upsert underlyingAssetIds if they don't exist in store
+                  underlyingAssetIds.forEach((underlyingAssetId, i) => {
+                    if (!assets[underlyingAssetId]) {
+                      const underlyingAsset = makeAsset({
+                        assetId: underlyingAssetId,
+                        symbol: asset.tokens[i].symbol,
+                        // No dice here, there's no name property
+                        name: asset.tokens[i].symbol,
+                        precision: asset.tokens[i].decimals,
+                        icon: asset.displayProps?.images[i] ?? '',
+                      })
+
+                      // TODO(gomes): we might want to do this in a batch
```

Comment:
let's do this now, upserting to assets will cause every selector to recompute

Reply:
https://github.com/shapeshift/web/pull/4504/commits/6e9a2e096bec2b52fdea6aba12b7958653b71007